### PR TITLE
🚨🚨 [Low Code CDK] Update `*ref` format to `#/`

### PR DESCRIPTION
**Pull request by clnoll**
Closes [19409](https://app.zenhub.com/workspaces/connector-extensibility-6257455decf3920012f4e872/issues/gh/airbytehq/airbyte/19409).

## What
Updates the reference format used in manifests from our custom format (`*ref(.*)`, with `.`-separated path components) to the format used by JSON Schema (`#/...`, with `/`-separated path components).

## How
Updates the manifest reference resolver to resolve references starting with `#/`.

Also includes a script to update all manifests to use the new reference format. When run, it will update 91 connectors to use the new format.

Note: with our original reference format, we were also handling references that index into a list element, which took a bracket notation (e.g. `*ref(definitions.mycomponent[0])`). In JSON Schema references, the brackets would be replaced with `/` (e.g. mycomponent/0). However, I didn't see any existing connectors using this feature so for the sake of simplicity haven't included it in the script.

## 🚨 User Impact 🚨
All manifests that use references will need to be updated to use the new reference format or else their references will not be resolved. The current thinking is that we will run the script and update the manifests, but won't publish a new version of the connector; the updated manifest will be present whenever the next update is ready for publishing.

### CODE REVIEW

File: airbyte-cdk/python/airbyte_cdk/sources/declarative/parsers/manifest_reference_resolver.py
Comment (sherifnada) [inline airbyte-cdk/python/airbyte_cdk/sources/declarative/parsers/manifest_reference_resolver.py:109]: just a q: why are we keeping backwards compatibility?

Comment (clnoll) [reply airbyte-cdk/python/airbyte_cdk/sources/declarative/parsers/manifest_reference_resolver.py:109]: We are no longer supporting the `*ref` format, so aren't keeping backwards compatibility there. The `$ref` format is supported by JSON Schema so we're still using it to identify refs whose value should be splatted into the referencer.

Comment (sherifnada) [reply airbyte-cdk/python/airbyte_cdk/sources/declarative/parsers/manifest_reference_resolver.py:109]: ah got it. Thanks for the 🥁 pointer 🥁

File: airbyte-cdk/python/airbyte_cdk/sources/declarative/parsers/manifest_reference_resolver.py
Comment (maxi297) [inline airbyte-cdk/python/airbyte_cdk/sources/declarative/parsers/manifest_reference_resolver.py:10]: Why extract this outside of the class boundaries? Shouldn't this be private to `ManifestReferenceResolver`?

**Review by maxi297 — APPROVED** [review]
LGTM! When will we migrate the manifest using the migration script?

File: airbyte-cdk/python/airbyte_cdk/sources/declarative/parsers/manifest_reference_resolver.py
Comment (clnoll) [reply airbyte-cdk/python/airbyte_cdk/sources/declarative/parsers/manifest_reference_resolver.py:10]: I could go either way. It allowed me to clean things up a little - making both `is_ref` and `is_ref_key` `staticmethod`s, plus I didn't feel strongly that it should be owned by `ManifestReferenceResolver`. That said, it is only _used_ by `ManifestReferenceResolver`, so I can see an argument for keeping it private.

**Comment by clnoll** [comment]
> When will we migrate the manifest using the migration script?

The migration will be done in a separate PR. Debating between waiting until other breaking changes are done and doing all of them in one pass to keep the connectors' commit histories cleaner, or running it just after we merge to `low_code_cdk_to_beta` so the commit histories of each connector explicitly state each breaking change that was applied to them. Leaning toward the latter, but WDYT?

cc @maxi297 @brianjlai

**Review by brianjlai — APPROVED** [review]
changes look good to me. nice work!

Have you run the `./gradlew validateSourceYamlManifest` command which verifies the `spec` command still works for all low code connectors? And it might be also worth testing a `read` on 1-2 connectors as a sanity check if you haven't already done so.

with regards to the migration, yeah I think I'm also leaning towards running the migration on the base `low_code_cdk_to_beta` branch. Just makes the reviews a little easier to manage and we can run the included script which can constitute individual commits as we look at the final branch

**Comment by brianjlai** [comment]
I'm also not sure what's going on w/ the Platform and Frontend CI checks. Strange

**Comment by clnoll** [comment]
@brianjlai just want to confirm that `./gradlew validateSourceYamlManifest` passes and the spot check of `read`s passed too.

**Comment by clnoll** [comment]
@maxi297 would you mind taking a look at my latest commit, [bdf1446](https://github.com/airbytehq/airbyte/pull/21434/commits/bdf1446f929df05cf7b9931312f181de85c8fd44)? I added it because I was getting a [build failure](https://github.com/airbytehq/airbyte/actions/runs/3962117292/jobs/6788427876#step:8:1024) with an error, `TS2305: Module '"core/request/ConnectorManifest"' has no exported member 'InterpolatedRequestOptionsProvider'.`, which I think is related to https://github.com/airbytehq/airbyte/issues/19410.

**Comment by brianjlai** [comment]
I'm thinking this is because the frontend hasn't been updated to reflect the now removed component since we got rid of it in the manifest. I think FE auto-generates their type definitions from the same manifest which is now missing that component. We need to file a ticket so Lake and Joe are aware, which should now be part of our acceptance criteria. But if we want to focus on building these faster, we can probably just ignore these tests when we merge to the base branch since we can tolerate the tests not working in the immediate until we update the front end code.

**Comment by clnoll** [comment]
Okay thanks @brianjlai in that case I'll revert [bdf1446](https://github.com/airbytehq/airbyte/pull/21434/commits/bdf1446f929df05cf7b9931312f181de85c8fd44).

**Pull request merged**
